### PR TITLE
Added health processor configuration serivce; Updated tests for ProcessorService

### PR DESCRIPTION
**Pull request by todorinskiz**

### CODE REVIEW

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+public interface HealthProcessorConfiguration {
+    boolean isEnabled();
+    boolean isRunOnce();
+    ProcessorService.Scope getScope();
+    int getTransactionLimit();
+    long getFirstTransaction();
+    int getTimeIncrementSeconds();
+    long getFirstCommitTime();
+    String getFirstCommitTimeValue();
```
Comment (driesdeproost) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfiguration.java:11]: Why is this a String and not a Date-like object?

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java [hunk 2]
```diff
@@ -12,23 +12,31 @@
...
+    @Getter @Setter
     protected RetryingTransactionHelper retryingTransactionHelper;
 
     @Getter @Setter
     protected ExecutorService executorService = Executors.newFixedThreadPool(5);
 
+    @Getter @Setter
+    protected HealthProcessorConfiguration configurationService;
+
     protected  ProcessorServiceImpl() {}
```
Comment (driesdeproost) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java:24]: Why is this constructor needed? Will it be used?
Can we just make this an `@Data` class with an `@AllArgsConstructor` and drop the separate `@Getter` and `@Setter` annotations and the explicit constructors?

**Review by driesdeproost — COMMENTED** [review]
build fails because alfresco-enterprise integration test subprojects are not present

I commented them out in settings.gradle locally to review.

I suppose we do not provide these subprojects since this is an open source project?
Should we disable enterprise by default and provide a mechanism to enable enterprise tests?

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java [hunk 3]
```diff
@@ -12,23 +12,31 @@
...
-    @Getter
-    @Setter
+    @Getter @Setter
     protected RetryingTransactionHelper retryingTransactionHelper;
 
     @Getter @Setter
     protected ExecutorService executorService = Executors.newFixedThreadPool(5);
 
+    @Getter @Setter
+    protected HealthProcessorConfiguration configurationService;
```
Comment (kerkhofsd) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java:22]: It's not really a service? `configuration` suffices?

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import lombok.Setter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Properties;
+
+public class HealthProcessorConfigurationImpl implements HealthProcessorConfiguration {
```
Comment (kerkhofsd) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java:14]: IMO the overhead of a separate interface is not necessary at this point. 
I also don't like the `...Impl` name of implementations. 

This article seems a good summary of my thoughts (especially paragraph 1.4): 
https://octoperf.com/blog/2016/10/27/impl-classes-are-evil/

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/model/PropertyConverter.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package eu.xenit.alfresco.processor.model;
+
+import eu.xenit.alfresco.processor.service.ProcessorService;
+import eu.xenit.alfresco.processor.service.ProcessorServiceImpl;
+
+import java.security.InvalidParameterException;
+
+public interface PropertyConverter<T> {
```
Comment (kerkhofsd) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/model/PropertyConverter.java:8]: Again, looks like unnecessary overhead to me.

Example: `PropertyConverter.BOOLEAN` is not shorter than `Boolean::parseBoolean` and even worse: requires an additional step / link to follow to get a grasp of what is actually going on.

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    public String getFirstCommitTimeValue() {
+        return getProperty(
+                PROP_PROC_TXN_COMMIT_TIME_START,
+                LocalDate.now().minusYears(10)
+                        .format(DateTimeFormatter.ISO_LOCAL_DATE),
+                PropertyConverter.STRING);
+    }
+
+    <T> T getProperty(String propertyName, T defaultValue, PropertyConverter<T> convertor) {
+        if(propertyName == null || propertyName.length() < 1) {
```
Comment (kerkhofsd) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java:102]: `.isEmpty()`?

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            throw new IllegalStateException("Property name must be provided in order to resolve a value.");
+        }
+
+        if(!globalProperties.containsKey(propertyName)) {
+            logger.warn("Property {} could not be found in the global configuration", propertyName);
```
Comment (kerkhofsd) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java:107]: `warn` is not the correct log severity, especially not if a default value is provided?

**Review by kerkhofsd — APPROVED** [review]
Approving since I have no blocking remarks. I do however have thoughts that can be discussed if necessary.

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java (hunk 4 above)
Comment (kerkhofsd) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java:14]: Or this: https://stackoverflow.com/a/2814831

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java (hunk 2 above)
Comment (tgeens) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java:24]: don't use `@Data` for services

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java [hunk 8]
```diff
@@ -1,5 +1,9 @@
 package eu.xenit.alfresco.processor.service;
 
 public interface ProcessorService {
+    enum Scope {
```
Comment (tgeens) [inline alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java:4]: what is `Scope` ?

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java (hunk 2 above)
Comment (tgeens) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java:24]: but :+1: for constructor injection and dropping unnecessary `@Getter` and `@Setter` annotations

**Comment by todorinskiz** [comment]
> build fails because alfresco-enterprise integration test subprojects are not present
> 
> I commented them out in settings.gradle locally to review.
> 
> I suppose we do not provide these subprojects since this is an open source project?
> Should we disable enterprise by default and provide a mechanism to enable enterprise tests?

Enterprise resources added, the tasks are enabled and repositories added when you use `-Penterprise`

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java (hunk 8 above)
Comment (todorinskiz) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java:4]: `TransactionScope` ? as in: transactions to scan :thinking:

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java (hunk 2 above)
Comment (todorinskiz) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java:24]: I'd still like to keep `@Setter(AccessLevel.PACKAGE)` to be able to only inject the mocked services?

Comment (todorinskiz) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java:24]: ... and `@NoArgsConstructor(access = AccessLevel.PACKAGE)` to be able to use `@Spy` ?

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java (hunk 7 above)
Comment (todorinskiz) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/HealthProcessorConfigurationImpl.java:107]: changed to `debug`

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java (hunk 2 above)
Comment (tgeens) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java]: 1. use constructor injection for the mocks
2. use `Mockito.spy(instance)` to wrap your instance with spy-instrumentation

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java (hunk 8 above)
Comment (tgeens) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java]: but we always scan ALL transactions, from start till the end - there is no *new* concept ?

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java (hunk 2 above)
Comment (todorinskiz) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorServiceImpl.java]: After discussion with Daan I already dropped these

File: alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java (hunk 8 above)
Comment (todorinskiz) [reply alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProcessorService.java]: Removed!

**Review by tgeens — APPROVED** [review]

**Pull request merged**